### PR TITLE
Update Enumeration

### PR DESCRIPTION
**Pull request by svick**
This removes unnecessary `new()` constraint, as suggested by https://github.com/dotnet/docs/issues/6812, synchronizes this code with dotnet/docs (see https://github.com/dotnet/docs/pull/6967) and also makes few other small improvements.

### CODE REVIEW

**Comment by svick** [comment]
Now that https://github.com/dotnet-architecture/eShopOnContainers/pull/653 is merged, this PR becomes smaller, but I think still useful.

I have also added a commit that partially reverts https://github.com/dotnet-architecture/eShopOnContainers/pull/653: I don't see any reason to create derived types of `CardType`, I think it's just boilerplate code that desn't add any value.

**Comment by CESARDELATORRE** [comment]
We'll be reviewing this PR in the upcoming days. Sorry, there are many other priorities.. :(
Thanks for the PR! 👍

**Comment by mairaw** [comment]
@CESARDELATORRE any updates here? This is blocking an open PR on docs as well:
https://github.com/dotnet/docs/pull/6967

**Comment by CESARDELATORRE** [comment]
@mvelosop - Can you check out this PR. I'm heads-down on other priorities and cannot make it now. Thanks! 👍

**Comment by mvelosop** [comment]
Working on this @svick, @CESARDELATORRE

File: src/Services/Ordering/Ordering.Domain/SeedWork/Enumeration.cs [hunk 1]
```diff
@@ -7,12 +7,9 @@ namespace Microsoft.eShopOnContainers.Services.Ordering.Domain.SeedWork
 {
     public abstract class Enumeration : IComparable
     {
-        public string Name { get; private set; }
+        public string Name { get; }
```
Comment (mvelosop) [inline src/Services/Ordering/Ordering.Domain/SeedWork/Enumeration.cs]:
```suggestion
        public string Name { get; private set; }
```

File: src/Services/Ordering/Ordering.Domain/SeedWork/Enumeration.cs [hunk 2]
```diff
@@ -7,12 +7,9 @@ namespace Microsoft.eShopOnContainers.Services.Ordering.Domain.SeedWork
...
 
-        public int Id { get; private set; }
-
-        protected Enumeration()
-        { }
+        public int Id { get; }
```
Comment (mvelosop) [inline src/Services/Ordering/Ordering.Domain/SeedWork/Enumeration.cs]:
```suggestion
        public int Id { get; private set; }
```

**Comment by ardalis** [comment]
Might be worth looking at something like this for faux enumerations?
https://github.com/ardalis/SmartEnum
Nuget: https://www.nuget.org/packages/Ardalis.SmartEnum

**Comment by mvelosop** [comment]
Hi @ardalis, thanks for the info 😊

I'm guessing @CESARDELATORRE was inspired by your post, for the code used here is just like the one there, and [the article in the guide](https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/microservice-ddd-cqrs-patterns/enumeration-classes-over-enum-types) has those references too!

However, for simplicity's sake I'd just leave it like this.

Actually, IMHO, I think using the Enumeration/SmartEnum class for CardType is not adecuate, since CardTypes are essentially "user-defined".

I find it's appropriate for [OrderStatus](https://github.com/dotnet-architecture/eShopOnContainers/blob/dev/src/Services/Ordering/Ordering.Domain/AggregatesModel/OrderAggregate/OrderStatus.cs) and the like, that's "domain-defined".

**Comment by varunkumar2310** [comment]
Can someone explain why we need "abstract" in class Enumeration.
It does not even have any abstract fields/ properties/ methods.

I wanted to create an default/ invalid enumeration for this class.
For example : 
public static Enumeration  Invalid = new Enumeration("Invalid", Int32.MinValue);

But c# compiler gives an error : 
cannot create instance of abstract class c# enumeration

**Comment by ardalis** [comment]
It's abstract to prevent anyone from creating an invalid version of the enumeration.

To do what you're trying to do, I would create a new `InvalidEnumeration` class that inherits from the abstract `Enumeration`. Then your code would work with the slight change of `new InvalidEnumeration`.

**Comment by varunkumar2310** [comment]
@ardalis 
Thank you :D

**Pull request closed without merging**